### PR TITLE
Add some spec best practices

### PR DESCRIPTION
**Pull request by thilonel**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -58,6 +58,40 @@ long_line_under_100_chars
 # Extra Specs Coding Style
 
 - Add a blank line between `let` and `before`.
+- Group specs around the described object's public methods. Use `describe "#method_name` for instance methods and `describe ".method_name"` for class methods.
+- Context setup does not belong to the `it` block, that's the place for the assert.
+- Context description should always start like `context "when ..."` or `context "and ..."`
```
Comment (Chambeur) [inline README.md]: Why not `context "with"`? 😄 
I think we cant use a less-restrictive style by allowing `grammatical relative clauses introduced by a pro-form` as a context. 
Or we keep this codding-style => http://www.betterspecs.org/#contexts

File: README.md [hunk 2]
```diff
@@ -58,6 +58,40 @@ long_line_under_100_chars
...
+- When you see `it "returns foo when baz is true"` it means you're missing a `context "when baz is true"`
```
Comment (Chambeur) [inline README.md]: Already referred here: http://www.betterspecs.org/#contexts

File: README.md [hunk 3]
```diff
@@ -58,6 +58,40 @@ long_line_under_100_chars
...
+- Prefer `build_stubbed` over `build` over `create` whenever possible. Helps keep spec suite fast, encourages DI, can prevent N+1 queries just by making you aware of DB queries.
+- When writing factories, don't specify more than what's needed to pass validation, put the more detailed stuff under well named traits.
+- Don't use random values in the factories, that can make specs flaky.
```
Comment (Chambeur) [inline README.md]: That means don't use `Faker`... 
I disagree with this rule as fixed value may create false negative.

**Comment by Chambeur** [comment]
As said on Slack:
I’m not a fan as when I read spec context, I usually don’t start from the subject. I’m more into following the same principle as for variables (the subject been the exception): (edited)
An object should be defined with objects that have been previously defined (ie: write first things that should be created first)
```ruby
let!(:needed_for_this) {...}
let(:this) { new(:this, needed_for_this) }
```
or with a real world example
```ruby
let!(:account) { create(:account) }
let(:rental) { build(:rental, account: account) }
```
In the case of multiple lines, if I read `let(:rental) { build(:rental, account: account) }` I’ll look for `account` definition upper in the file (not lower).

Also in the case of multiple `let` and `before` (I think it’s a smell personally), it’s easier for the brain to group which one are used with the others:
```ruby
let!(:needed_for_this) {...}
let(:this) { new(:this, needed_for_this) }
let(:this_stuff) { new(:stuff) }
before { this.do(this_stuff) }

let!(:needed_for_that) {...}
let(:that) { new(:that, needed_for_that) }
let(:that_stuff) { new(:stuff) }
before { that.do(that_stuff) }
```

Doing this, it's simplier to extract `context`:
```ruby
context "when this" do
  let!(:needed_for_this) {...}
  let(:this) { new(:this, needed_for_this) }
  let(:this_stuff) { new(:stuff) }
  before { this.do(this_stuff) }
end

context "when that" do
  let!(:needed_for_that) {...}
  let(:that) { new(:that, needed_for_that) }
  let(:that_stuff) { new(:stuff) }
  before { that.do(that_stuff) }
end
```

IMO, `let` and `let!` differences and the calling process are irrelevant for the code readability as it’s designed for test suite execution speed. So I think the common rules of thumbs are more appropriate:
- declare your variable just before where you use it
- group the code lines that belongs together

File: README.md (hunk 1 above)
Comment (thilonel) [reply README.md]: Right, `with` is okay too. I don't understand the second sentence.

File: README.md (hunk 2 above)
Comment (thilonel) [reply README.md]: True, but I guess it doesn't hurt to emphasize 😄 Shall I remove this?

File: README.md (hunk 3 above)
Comment (thilonel) [reply README.md]: Yeah, we shouldn't use faker in the factory. It is very annoying when a spec fails on CI but you don't know what it's input value was and you can't reproduce it locally.
Faker is good for demo data, I don't see how it's necessary for tests.

> I disagree with this rule as fixed value may create false negative.

What does this mean? You lost me at false negative. 😅

File: README.md (hunk 1 above)
Comment (Chambeur) [reply README.md]: I was trying to say that the rule is too restrictive when taken literally. :)
I think it's also redundant with this: http://www.betterspecs.org/#contexts

File: README.md (hunk 2 above)
Comment (Chambeur) [reply README.md]: When there is too much to read, people tend to read less.
That's why I would remove it.

File: README.md (hunk 3 above)
Comment (Chambeur) [reply README.md]: I mean't I disagree with the `Don't use random value` rule.
Because flaky test is not bad itself (it's just very annoying), it means that either the code doesn't handle an edge case, either the spec setup is wrong and then should be fixed.

I know this opinion is not shared with everyone so I would not add the `Don't use random value` rule until everyone agree or disagree.

**Comment by thilonel** [comment]
@Chambeur I think the rest I addressed on Slack. Are we good?
